File: ios/App/UI/AppSettings.swift
import Foundation
import PhotocellCore

/// Configurações persistidas (UserDefaults). Os padrões seguem a especificação e as decisões do estudo.
struct AppSettings: Codable, Equatable {
    var stripWidthPx: Int = 15
    var coreWidth: Int = 3
    var lineXFraction: Double = 0.5          // posição da linha na tela (fração da largura do preview)
    var bandTopFraction: Double = 0.25       // banda de linhas (fração da altura do preview)
    var bandBottomFraction: Double = 0.75
    var exposureNs: Int64 = 2_083_333        // 1/480 s
    var startLockoutMs: Int = 1500
    var frameResumeS: Double = 8.0
    var finishArmS: Double = 10.0
    var finishLockoutMs: Int = 2000
    var confirmRequired: Int = 2
    var confirmWindow: Int = 4
    var flickerAuto: Bool = true
    var suspendStrategy: SuspendStrategy = .disableConnection
    var skewNs: Int64? = nil
    var thresholdFloor: Double = 4.0
    var feedbackSound: Bool = true
    var feedbackFlash: Bool = true

    static let exposureChoices: [(label: String, ns: Int64)] = [
        ("1/240 s (sem janela cega, mais blur)", 4_166_666),
        ("1/480 s (padrão)", 2_083_333),
        ("1/1000 s", 1_000_000),
        ("1/2000 s (sol forte)", 500_000),
        ("1/4000 s", 250_000),
    ]

    func makeConfig() -> PhotocellConfig {
        var c = PhotocellConfig()
        c.startLockoutNs = Int64(startLockoutMs) * 1_000_000
        c.frameResumeNs = Int64(frameResumeS * 1e9)
        c.finishArmNs = Int64(finishArmS * 1e9)
        c.finishLockoutNs = Int64(finishLockoutMs) * 1_000_000
        c.confirmRequired = confirmRequired
        c.confirmWindow = confirmWindow
        c.flickerAuto = flickerAuto
        c.coreWidth = coreWidth
        c.exposureNs = exposureNs
        c.skewNs = skewNs
        c.thresholdFloor = thresholdFloor
        return c
    }

    private static let key = "fotocelula.settings.v1"

    static func load() -> AppSettings {
        guard let data = UserDefaults.standard.data(forKey: key),
              let s = try? JSONDecoder().decode(AppSettings.self, from: data) else { return AppSettings() }
        return s
    }

    func save() {
        if let data = try? JSONEncoder().encode(self) {
            UserDefaults.standard.set(data, forKey: Self.key)
        }
    }
}

File: android/app/src/main/kotlin/br/com/tportooliveira/fotocelula/ui/AppSettings.kt
package br.com.tportooliveira.fotocelula.ui

import android.content.Context
import br.com.tportooliveira.fotocelula.core.PhotocellConfig

/** Configurações persistidas (SharedPreferences). Padrões = especificação + decisões do estudo. */
data class AppSettings(
    val stripWidthPx: Int = 15,
    val coreWidth: Int = 3,
    val lineXFraction: Float = 0.5f,
    val bandTopFraction: Float = 0.25f,
    val bandBottomFraction: Float = 0.75f,
    val exposureNs: Long = 2_083_333L,
    val startLockoutMs: Int = 1500,
    val frameResumeS: Float = 8.0f,
    val finishArmS: Float = 10.0f,
    val finishLockoutMs: Int = 2000,
    val confirmRequired: Int = 2,
    val flickerAuto: Boolean = true,
    val feedbackSound: Boolean = true,
    val feedbackFlash: Boolean = true,
) {
    fun makeConfig(fps: Int, appliedExposureNs: Long): PhotocellConfig = PhotocellConfig(
        frameRateHz = fps,
        startLockoutNs = startLockoutMs * 1_000_000L,
        frameResumeNs = (frameResumeS * 1e9).toLong(),
        finishArmNs = (finishArmS * 1e9).toLong(),
        finishLockoutNs = finishLockoutMs * 1_000_000L,
        confirmRequired = confirmRequired,
        flickerAuto = flickerAuto,
        coreWidth = coreWidth,
        exposureNs = if (appliedExposureNs > 0) appliedExposureNs else exposureNs,
        calibrationSamples = fps,          // 1 s de amostras em qualquer taxa
    )

    companion object {
        val exposureChoices = listOf(
            "1/240 s (sem janela cega, mais blur)" to 4_166_666L,
            "1/480 s (padrão)" to 2_083_333L,
            "1/1000 s" to 1_000_000L,
            "1/2000 s (sol forte)" to 500_000L,
            "1/4000 s" to 250_000L,
        )

        fun load(ctx: Context): AppSettings {
            val p = ctx.getSharedPreferences("fotocelula", Context.MODE_PRIVATE)
            val d = AppSettings()
            return AppSettings(
                stripWidthPx = p.getInt("stripWidthPx", d.stripWidthPx),
                coreWidth = p.getInt("coreWidth", d.coreWidth),
                lineXFraction = p.getFloat("lineX", d.lineXFraction),
                bandTopFraction = p.getFloat("bandTop", d.bandTopFraction),
                bandBottomFraction = p.getFloat("bandBottom", d.bandBottomFraction),
                exposureNs = p.getLong("exposureNs", d.exposureNs),
                startLockoutMs = p.getInt("startLockoutMs", d.startLockoutMs),
                frameResumeS = p.getFloat("frameResumeS", d.frameResumeS),
                finishArmS = p.getFloat("finishArmS", d.finishArmS),
                finishLockoutMs = p.getInt("finishLockoutMs", d.finishLockoutMs),
                confirmRequired = p.getInt("confirmRequired", d.confirmRequired),
                flickerAuto = p.getBoolean("flickerAuto", d.flickerAuto),
                feedbackSound = p.getBoolean("feedbackSound", d.feedbackSound),
                feedbackFlash = p.getBoolean("feedbackFlash", d.feedbackFlash),
            )
        }
    }

    fun save(ctx: Context) {
        ctx.getSharedPreferences("fotocelula", Context.MODE_PRIVATE).edit()
            .putInt("stripWidthPx", stripWidthPx).putInt("coreWidth", coreWidth)
            .putFloat("lineX", lineXFraction).putFloat("bandTop", bandTopFraction).putFloat("bandBottom", bandBottomFraction)
            .putLong("exposureNs", exposureNs).putInt("startLockoutMs", startLockoutMs)
            .putFloat("frameResumeS", frameResumeS).putFloat("finishArmS", finishArmS).putInt("finishLockoutMs", finishLockoutMs)
            .putInt("confirmRequired", confirmRequired).putBoolean("flickerAuto", flickerAuto)
            .putBoolean("feedbackSound", feedbackSound).putBoolean("feedbackFlash", feedbackFlash)
            .apply()
    }
}
